pkg/gencosmosdb/cosmosdb/template_fake.go

```diff
@@ -0,0 +1,301 @@
+	}
+
+	if options != nil {
+		for _, o := range options.PreTriggers {
+			err := c.processPreTrigger(ctx, doc, o)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	buf := &bytes.Buffer{}
+	err := codec.NewEncoder(buf, c.jsonHandle).Encode(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	out := buf.String()
+	c.docs[doc.ID] = &out
+	return c.fromString(&out), nil
+}
+
+func (c *FakeTemplateClient) Delete(ctx context.Context, partitionKey string, doc *pkg.Template, options *Options) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	_, ext := c.docs[doc.ID]
+	if !ext {
+		return &Error{StatusCode: http.StatusNotFound}
+	}
+
+	delete(c.docs, doc.ID)
+	return nil
+}
+
+func (c *FakeTemplateClient) ChangeFeed(*Options) TemplateIterator {
+	return &fakeTemplateNotImplementedIterator{}
+}
+
+func (c *FakeTemplateClient) processPreTrigger(ctx context.Context, doc *pkg.Template, trigger string) (err error) {
+	trig, ok := c.triggers[trigger]
+	if ok {
+		return trig(ctx, doc)
+	} else {
+		panic(ErrNotImplemented)
+	}
+}
+
+func (c *FakeTemplateClient) Query(name string, query *Query, options *Options) TemplateRawIterator {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	quer, ok := c.queries[query.Query]
+	if ok {
+		docs := quer(query, c.docs, c.fromString)
+		return &FakeTemplateClientRawIterator{
+			docs:       docs,
+			jsonHandle: c.jsonHandle,
+		}
+	} else {
+		panic(ErrNotImplemented)
+	}
+}
+
+func (c *FakeTemplateClient) QueryAll(ctx context.Context, partitionkey string, query *Query, options *Options) (*pkg.Templates, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	var results []*pkg.Template
+
+	quer, ok := c.queries[query.Query]
+	if ok {
+		for _, doc := range quer(query, c.docs, c.fromString) {
+			results = append(results, c.fromString(doc))
+		}
+	} else {
+		panic(ErrNotImplemented)
+	}
+
+	return &pkg.Templates{
+		Count:     len(results),
+		Templates: results,
+	}, nil
+}
+
+func (c *FakeTemplateClient) InjectTrigger(trigger string, impl FakeTemplateTrigger) {
+	c.triggers[trigger] = impl
+}
+
+func (c *FakeTemplateClient) InjectQuery(query string, impl FakeTemplateQuery) {
+	c.queries[query] = impl
+}
+
+type FakeTemplateClientRawIterator struct {
+	docs         []*string
+	jsonHandle   *codec.JsonHandle
+	continuation int
+}
+
+func (i *FakeTemplateClientRawIterator) decode(inp []*string) (done []*pkg.Template, err error) {
+	for _, doc := range inp {
+		res := &pkg.Template{}
+		d := codec.NewDecoder(bytes.NewBufferString(*doc), i.jsonHandle)
+		err := d.Decode(&res)
+		if err != nil {
+			return nil, err
+		}
+		done = append(done, res)
+	}
+	return
+}
+
+func (i *FakeTemplateClientRawIterator) Next(ctx context.Context, maxItemCount int) (*pkg.Templates, error) {
+	docs := &pkg.Templates{}
+	err := i.NextRaw(ctx, maxItemCount, docs)
+	return docs, err
+}
+
+func (i *FakeTemplateClientRawIterator) NextRaw(ctx context.Context, maxItemCount int, raw interface{}) (err error) {
```

Comment:
some but not all of the logic in this function is going to need to stick around, but redirect it to something like (struct deliberately not exported) `func (i fakeTemplateClientListIterator) Next(...)`